docs/Execution.md

```diff
@@ -0,0 +1,107 @@
+Execution
+======
+
+The MsQuic API uses a different execution model compared to BSD-style sockets and most other networking libraries built on them.
+The sections below detail the designs MsQuic uses and the reasons behind these choices.
+
+## Event Model
+
+In the MsQuic API, all state changes and notifications are indicated directly to the application via a callback.
+This includes connection state changes, new streams being created, stream data being received, and stream sends completing.
+
+```c
+typedef struct QUIC_LISTENER_EVENT {
+    QUIC_LISTENER_EVENT_TYPE Type;
+    union {
+        struct { ... } NEW_CONNECTION;
+        struct { ... } STOP_COMPLETE;
+    };
+} QUIC_LISTENER_EVENT;
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_LISTENER_CALLBACK)
+QUIC_STATUS
+(QUIC_API QUIC_LISTENER_CALLBACK)(
+    _In_ HQUIC Listener,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_LISTENER_EVENT* Event
+    );
+```
+
+Above is an example of a callback delivered to the listener interface.
+The application must register a callback handler to manage all events MsQuic may indicate, returning a status to show if it was successfully handled or not.
+
+This approach differs significantly from BSD sockets, where the application must make a call (e.g., `send` or `recv`) to determine if something happened.
+This design choice was made for several reasons:
+
+- The MsQuic API **runs in-process**, eliminating the need for a kernel to user mode boundary switch to notify the application layer. This makes the callback-based design more practical compared to BSD sockets.
+
+- MsQuic, due to the QUIC protocol, has numerous event types. Applications may have hundreds of objects with potential state changes. The callback model allows the application to avoid managing pending calls on each object.
+
+- Writing correct, performant code on top of the BSD-style interface has proven challenging. Callbacks, executed at the correct time and on the correct thread/processor, enable MsQuic to abstract much complexity from applications, making things "just work" out of the box.
```

Comment:
I'm not a big fan of this statement. one can write correct code using select, and on non-Windows operating systems, select performs well enough for many (most?) apps in the past. it's just Windows did not optimize the non-blocking IO path like *nix OS's did.

This statement also sounds non-convincing:  'making things "just work" out of the box". When I read through this, I don't see it "just work" - I just see my business logic moving from select() loops to callbacks.

That change *can* be a good thing - it avoids burning threads when one must poll with select calls. so I could see the argument made that these callbacks *scale* much more effectively because one doesn't need to manually poll against X number of sockets on Y number of threads -- scaling *out* to large numbers of sockets can indeed be problematic.

This is where OVERLAPPED I/O helps scale better than non-blocking I/O - with very large #s of sockets.

It would be useful to differentiate:
how does this callback model change from the non-blocking model for the benefit of the developer?
IMHO, it helps by (a) scaling out naturally without the developer creating their own threadpool and eventing workflow (e.g., avoiding thundering herd problems).

How does this callback model change from the OVERLAPPED I/O model for the benefit of the developer?
IMHO, it helps by not requiring posting an I/O to just receive state change notifications; it helps by not requiring the developer to post I/O up-front, which has a scale-out problem (if I have 100k connections, I don't want to have to pend 100K buffer allocations - I just want to know when a connection is ready, so I can post a send/recv with my buffer when the socket is ready to handle the IO)

Reply:
Perhaps we can agree to disagree? 😆 

I have been working on our socket abstraction for all the platforms for over half a decade and I'd argue it's still not perfect (on any platform). So, IMHO, it is not possible to get sockets working out of the box. Even with the years of dedicated networking experience, and the fact that we **own** the Windows layers, we still struggle to get it right. We also get near constant internal partner questions/investigations that lead me towards the conclusion that others have similar problems.

That being said, I'm totally fine with making text changes here.